### ISSUE TITLE
FileNotFoundException while the file exists and is available

### ISSUE THREAD

**Issue by zandervandermeer** [thephpleague/flysystem#741]
Hi everyone,

We are having some trouble using the Flysystem package in combination with the flysystem-rackspace (https://github.com/thephpleague/flysystem-rackspace) and flysystem-laravel (https://github.com/GrahamCampbell/Laravel-Flysystem) packages. I'm hoping some of you may be able to shed some light onto this issue and help us.

**The problem:**
Our Laravel application is unable to retreive a few specific images from Rackspace using the filesystem manager and will result in a `League\Flysystem\FileNotFoundException` while the image exists and is available. (This has been verified using Cyberduck and the online browser)

So far this only happend when accessing the API using a job and never when doing it inside the request cycle. However it maybe possible that this is purely coincidence. 

**Code:**
To try and collect some additional information I've added a try/catch block that includes my own additional data, however this didn't work as expected because the catch is never executed.
(Code block starts at line 66)
```php
try {
    $binary = $this->fileSystem()->get($media->path . '/' . $media->filename);
} catch (\Exception $e) {
    throw new FileUploadException('Failed to retrieve uploaded file for resizing', [
        'Media ID'  => $media->id,
        'Model'     => is_object($model) ? $model->id : $model,
        'Width'     => $width,
        'Height'    => $height,
        'Ratio'     => $ratio,
    ] + (array) $e);
}

protected function fileSystem()
{
    if (!isset($this->fileSystem)) {
        $this->fileSystem = app('filesystem')->disk();
    }
    return $this->fileSystem;
}
```

**Stacktrace:**
```[2017-01-05 11:09:20] production.ERROR: League\Flysystem\FileNotFoundException: File not found at path: image/content/536x536/94fad6d1ac5883a936c5aa84857ea143.jpg in /home/forge/example.com/releases/20161117090927/vendor/league/flysystem/src/Filesystem.php:385
Stack trace:
#0 /home/forge/example.com/releases/20161117090927/vendor/league/flysystem/src/Filesystem.php(178): League\Flysystem\Filesystem->assertPresent('image/content/5...')
#1 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Filesystem/FilesystemAdapter.php(56): League\Flysystem\Filesystem->read('image/content/5...')
#2 /home/forge/example.com/releases/20161117090927/app/Services/Management/Image/ImageUploader.php(67): Illuminate\Filesystem\FilesystemAdapter->get('image/content/5...')
#3 /home/forge/example.com/releases/20161117090927/app/Jobs/Management/ResizeImagesFromMedia.php(53): App\Services\Management\Image\ImageUploader->resizedUpload(Object(App\Models\Media), Object(App\Models\Content), '128', '128', false)
#4 [internal function]: App\Jobs\Management\ResizeImagesFromMedia->handle()
#5 /home/forge/example.com/releases/20161117090927/bootstrap/cache/compiled.php(1187): call_user_func_array(Array, Array)
#6 /home/forge/example.com/releases/20161117090927/bootstrap/cache/compiled.php(9491): Illuminate\Container\Container->call(Array)
#7 [internal function]: Illuminate\Bus\Dispatcher->Illuminate\Bus\{closure}(Object(App\Jobs\Management\ResizeImagesFromMedia))
#8 /home/forge/example.com/releases/20161117090927/bootstrap/cache/compiled.php(9643): call_user_func(Object(Closure), Object(App\Jobs\Management\ResizeImagesFromMedia))
#9 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(App\Jobs\Management\ResizeImagesFromMedia))
#10 /home/forge/example.com/releases/20161117090927/bootstrap/cache/compiled.php(9625): call_user_func(Object(Closure), Object(App\Jobs\Management\ResizeImagesFromMedia))
#11 /home/forge/example.com/releases/20161117090927/bootstrap/cache/compiled.php(9498): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#12 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php(43): Illuminate\Bus\Dispatcher->dispatchNow(Object(App\Jobs\Management\ResizeImagesFromMedia), Object(Closure))
#13 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Queue/Jobs/Job.php(129): Illuminate\Queue\CallQueuedHandler->call(Object(Illuminate\Queue\Jobs\DatabaseJob), Array)
#14 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Queue/Jobs/DatabaseJob.php(50): Illuminate\Queue\Jobs\Job->resolveAndFire(Array)
#15 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(218): Illuminate\Queue\Jobs\DatabaseJob->fire()
#16 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(159): Illuminate\Queue\Worker->process('database', Object(Illuminate\Queue\Jobs\DatabaseJob), '2', 0)
#17 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(111): Illuminate\Queue\Worker->pop('database', 'default', 0, '10', '2')
#18 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(85): Illuminate\Queue\Worker->runNextJobForDaemon('database', 'default', 0, '10', '2')
#19 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(102): Illuminate\Queue\Worker->daemon('database', 'default', 0, 128, '10', '2')
#20 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(70): Illuminate\Queue\Console\WorkCommand->runWorker('database', 'default', 0, 128, true)
#21 [internal function]: Illuminate\Queue\Console\WorkCommand->fire()
#22 /home/forge/example.com/releases/20161117090927/bootstrap/cache/compiled.php(1187): call_user_func_array(Array, Array)
#23 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Console/Command.php(150): Illuminate\Container\Container->call(Array)
#24 /home/forge/example.com/releases/20161117090927/vendor/symfony/console/Command/Command.php(256): Illuminate\Console\Command->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#25 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Console/Command.php(136): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#26 /home/forge/example.com/releases/20161117090927/vendor/symfony/console/Application.php(846): Illuminate\Console\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#27 /home/forge/example.com/releases/20161117090927/vendor/symfony/console/Application.php(190): Symfony\Component\Console\Application->doRunCommand(Object(Illuminate\Queue\Console\WorkCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#28 /home/forge/example.com/releases/20161117090927/vendor/symfony/console/Application.php(121): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#29 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(107): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#30 /home/forge/example.com/releases/20161117090927/artisan(35): Illuminate\Foundation\Console\Kernel->handle(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#31 {main}

Next Illuminate\Contracts\Filesystem\FileNotFoundException: image/content/536x536/94fad6d1ac5883a936c5aa84857ea143.jpg in /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Filesystem/FilesystemAdapter.php:58
Stack trace:
#0 /home/forge/example.com/releases/20161117090927/app/Services/Management/Image/ImageUploader.php(67): Illuminate\Filesystem\FilesystemAdapter->get('image/content/5...')
#1 /home/forge/example.com/releases/20161117090927/app/Jobs/Management/ResizeImagesFromMedia.php(53): App\Services\Management\Image\ImageUploader->resizedUpload(Object(App\Models\Media), Object(App\Models\Content), '128', '128', false)
#2 [internal function]: App\Jobs\Management\ResizeImagesFromMedia->handle()
#3 /home/forge/example.com/releases/20161117090927/bootstrap/cache/compiled.php(1187): call_user_func_array(Array, Array)
#4 /home/forge/example.com/releases/20161117090927/bootstrap/cache/compiled.php(9491): Illuminate\Container\Container->call(Array)
#5 [internal function]: Illuminate\Bus\Dispatcher->Illuminate\Bus\{closure}(Object(App\Jobs\Management\ResizeImagesFromMedia))
#6 /home/forge/example.com/releases/20161117090927/bootstrap/cache/compiled.php(9643): call_user_func(Object(Closure), Object(App\Jobs\Management\ResizeImagesFromMedia))
#7 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(App\Jobs\Management\ResizeImagesFromMedia))
#8 /home/forge/example.com/releases/20161117090927/bootstrap/cache/compiled.php(9625): call_user_func(Object(Closure), Object(App\Jobs\Management\ResizeImagesFromMedia))
#9 /home/forge/example.com/releases/20161117090927/bootstrap/cache/compiled.php(9498): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#10 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php(43): Illuminate\Bus\Dispatcher->dispatchNow(Object(App\Jobs\Management\ResizeImagesFromMedia), Object(Closure))
#11 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Queue/Jobs/Job.php(129): Illuminate\Queue\CallQueuedHandler->call(Object(Illuminate\Queue\Jobs\DatabaseJob), Array)
#12 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Queue/Jobs/DatabaseJob.php(50): Illuminate\Queue\Jobs\Job->resolveAndFire(Array)
#13 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(218): Illuminate\Queue\Jobs\DatabaseJob->fire()
#14 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(159): Illuminate\Queue\Worker->process('database', Object(Illuminate\Queue\Jobs\DatabaseJob), '2', 0)
#15 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(111): Illuminate\Queue\Worker->pop('database', 'default', 0, '10', '2')
#16 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(85): Illuminate\Queue\Worker->runNextJobForDaemon('database', 'default', 0, '10', '2')
#17 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(102): Illuminate\Queue\Worker->daemon('database', 'default', 0, 128, '10', '2')
#18 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(70): Illuminate\Queue\Console\WorkCommand->runWorker('database', 'default', 0, 128, true)
#19 [internal function]: Illuminate\Queue\Console\WorkCommand->fire()
#20 /home/forge/example.com/releases/20161117090927/bootstrap/cache/compiled.php(1187): call_user_func_array(Array, Array)
#21 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Console/Command.php(150): Illuminate\Container\Container->call(Array)
#22 /home/forge/example.com/releases/20161117090927/vendor/symfony/console/Command/Command.php(256): Illuminate\Console\Command->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#23 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Console/Command.php(136): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#24 /home/forge/example.com/releases/20161117090927/vendor/symfony/console/Application.php(846): Illuminate\Console\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#25 /home/forge/example.com/releases/20161117090927/vendor/symfony/console/Application.php(190): Symfony\Component\Console\Application->doRunCommand(Object(Illuminate\Queue\Console\WorkCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#26 /home/forge/example.com/releases/20161117090927/vendor/symfony/console/Application.php(121): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#27 /home/forge/example.com/releases/20161117090927/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(107): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#28 /home/forge/example.com/releases/20161117090927/artisan(35): Illuminate\Foundation\Console\Kernel->handle(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#29 {main}```

**Comment by frankdejonge** [comment]
@zandervandermeer are you sure this code is related to this problem? Because that catch statement should catch is because it is the base exception type. The case you're describing would hint to an extremely unlikely php bug.

**Comment by zandervandermeer** [comment]
@frankdejonge We are using Bugsnag to log exceptions and I tried to add as much data to my own exceptions so finding the issue would be bit easier. But since my own catch is not being executed I do not have any additional data. That is the reason why I mentioned it here. So no, I'm not sure if it is related to this issue, but I think it is (was) important to mention. 

Do you have a bit more information about the 'extremely unlikely php bug' so I can do some digging on my own?

Thank you for your reply :)

**Comment by frankdejonge** [comment]
@zandervandermeer with "extremely unlikely" I actually meant the following: Because the catch block is not executed you can quite safely assume that the code in the code sample is in fact not the code causing the exception. The catch block looks file and should catch any exceptions.

**Comment by zandervandermeer** [comment]
@frankdejonge Thank you for your reply, however reading the stacktrace tells me it is in fact the code sample I provided. 

This line from the stacktrace is saying that the get() function from the FilesystemAdapter is throwing an exception, called from line 67 of the ImageUploader class. 
```
#2 /home/forge/example.com/releases/20161117090927/app/Services/Management/Image/ImageUploader.php(67): Illuminate\Filesystem\FilesystemAdapter->get('image/content/5...')
```

Line 67 of the ImageUploader class:
```php
$binary = $this->fileSystem()->get($media->path . '/' . $media->filename);
```
The exact line that is included in my code sample.

Am I reading the stacktrace wrong or is something else happening?

**Comment by ryanlovett-amb** [comment]
I encountered this same issue.  FWIW here was my experience.

My tracing indicated a user permissions issue with the worker.  Inside the request cycle everything was executed as www-data (nginx on Ubuntu 16.04). I had configured supervisor to run the queue watcher as root and, lo and behold this same error (albeit with a dropbox flysystem adaptor) manifested.

I changed the user in supervisor so that the queue watcher ran as www-data and the issue immediately resolved.

To me this makes no sense, but I hope my experience may help someone else.

**Comment by skerric** [comment]
I encountered this same issue. FWIW here was my experience.

My tracing indicated a user permissions issue with the worker. Inside the request cycle everything was executed as www-data (nginx on Ubuntu 16.04). I had configured supervisor to run the queue watcher as root and, lo and behold this same error (albeit with a dropbox flysystem adaptor) manifested.

I changed the user in supervisor so that the queue watcher ran as www-data and the issue immediately resolved.

To me this makes no sense, but I hope my experience may help someone else.
